native/cocos/audio/apple/AudioEngine-impl.mm

```diff
@@ -0,0 +1,336 @@
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#import "AVFoundation/AVAudioEngine.h"
+#import "AVFoundation/AVAudioPlayerNode.h"
+#import "AVFoundation/AVAudioFile.h"
+#import "AVFoundation/AVAudioFormat.h"
+#include "audio/apple/AudioEngine-Impl.h"
+#include "audio/include/AudioEngine.h"
+#include "application/ApplicationManager.h"
+#include "platform/FileUtils.h"
+
+static AVAudioEngine* engine_instance = nullptr;
+static AVAudioFormat* uniform_format = nullptr;
+namespace cc {
+static AudioEngineImpl* s_instance = nullptr;
+class Scheduler;
+
+
+AudioEngineImpl::AudioEngineImpl():_currentID(0), _lazyInitLoop(true){
+    engine_instance = [[AVAudioEngine alloc] init];
+    s_instance = this;
+}
+
+AudioEngineImpl::~AudioEngineImpl() {
+    if (auto sche = _scheduler.lock()) {
+        sche->unschedule("AudioEngine", this);
+    }
+    
+    //TODO: release all player nodes attached?
+    [engine_instance release];
+    _players.clear();
+    _caches.clear();
+    s_instance = nullptr;
+}
+
+bool AudioEngineImpl::init() {
+    _scheduler = CC_CURRENT_ENGINE()->getScheduler();
+    return true;
+    
+}
+
+// Return player in memo pool
+int32_t AudioEngineImpl::getUsablePlayer() {
+    if(_unusedPlayers.empty()) {
+        return -1;
+    }
+    return _unusedPlayers.begin()->first;
+}
+AudioCache* AudioEngineImpl::preload(const ccstd::string &filePath, const LoadCallback &callback) {
+    AudioCache* cache;
+    auto itr = _caches.find(filePath);
+    
+    if(itr != _caches.end()){
+        cache = itr->second;
+    } else {
+        auto fileFullPath = FileUtils::getInstance()->fullPathForFilename(filePath);
+        cache = new AudioCache(fileFullPath);
+        _caches[filePath] = cache;
+        AudioEngine::addTask([cache](){
+            cache->load();
+        });
+        _caches[filePath] = cache;
+    }
+    if(cache && callback) {
+        cache->addLoadCallback(callback);
+    }
+    return cache;
+}
+
+AudioCache* AudioEngineImpl::forceLoad(const ccstd::string &filePath, const LoadCallback &callback) {
+    AudioCache* cache;
+    auto itr = _caches.find(filePath);
+    
+    if(itr != _caches.end()){
+        cache = itr->second;
+        cache->load(); // force load async
+        if(callback) {
+            cache->addLoadCallback(callback);
+        }
+    } else {
+        auto fileFullPath = FileUtils::getInstance()->fullPathForFilename(filePath);
+        cache = new AudioCache(fileFullPath);
+        _caches[filePath] = cache;
+        cache->load(); // force load
+        _caches[filePath] = cache;
+        if(callback) {
+            cache->addLoadCallback(callback);
+        }
+    }
+    return cache;
+}
+
+int32_t AudioEngineImpl::registerAudio() {
+    AudioPlayer* player;
+    int32_t audioID = getUsablePlayer();
```

Comment:
Does it mean `playerId` here?  And the `_currentID` is increased only when _unusedPlayers is empty, it not related to audioId.

Reply:
`_unusedPlayers` is more sutable as list not a map